core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -697,6 +714,31 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
         info(s"Rejected connection from ${e.ip}, address already has the configured maximum of ${e.count} connections.")
         close(endPoint.listenerName, socketChannel)
         None
+      case e: ConnectionThrottledException =>
+        val ip = socketChannel.socket.getInetAddress
+        debug(s"Delaying closing of connection from $ip for ${e.throttleTimeMs} ms")
+        val delayQueue = throttledSockets.computeIfAbsent(ip, _ => new mutable.Queue[DelayedCloseSocket])
```

Comment:
I went with this data structure because throttle end time should be monotonic for a specific IP but not necessarily for all IPs. By maintaining a queue per IP, we can get constant time adds/removes for a throttled IP.

The downside of this is that in the case where there are no connections to unthrottle, `closeThrottledConnections` runs in `O(n)` for number of throttled IPs.

On the other hand `DelayQueue` uses a heap internally and has `O(log(n))` for each add/remove, and constant time for checking whether there are IPs to unthrottle. In scenarios where many connections will get unthrottled at once, the `DelayQueue` will perform significantly worse, since each `DelayQueue.dequeue` call will be `O(log(n))`.

I admit this could be a premature optimization, and a delay queue could work better in practice. Let me know what you think.

Reply:
I agree, that it is better to optimize for checking connections vs. overhead of adding/removing to the queue, because `closeThrottledConnections` runs pretty often (on a loop), which I think also means that finding that there are no connections yet to unthrottle would also be common. Or very few connections to unthrottle. So, after reading all your evaluations above, I am also leaning towards using a delay queue here.

Not sure if the question #1 about why we need to delay closing a connection got answered, so answering just in case. Since we want to throttle accepting connections from an IP, closing a connection due to reaching IP quota right away would not help with throttling accepting connections since that IP is going to reconnect. 

if we close the connection right away, that IP is going to reconnect right away, adn

Reply:
Okay, I replaced the map/queue with a priority queue.
I did not use the java `DelayQueue` similar to the implementation in `ClientQuotaManager` because this throttling implementation does not use timeout-based polling or require a synchronized data structure, and there's significantly more boilerplate needed for a `DelayQueue`